src/dataflow/source/kafka.rs

```diff
@@ -652,135 +702,117 @@ where
-                                bytes_read += key.len() as i64;
-                                bytes_read += out.len() as i64;
-                                let ts_cap = cap.delayed(&ts);
-                                output.session(&ts_cap).give(SourceOutput::new(
-                                    key,
-                                    out,
-                                    Some(Into::<KafkaOffset>::into(offset).offset),
-                                ));
-
-                                partition_metrics.offset_ingested.set(offset.offset);
-                                partition_metrics.messages_ingested.inc();
-
-                                downgrade_capability(
-                                    &id,
-                                    cap,
-                                    &mut cp_info,
-                                    &mut dp_info,
-                                    &timestamp_histories,
-                                );
-                            }
-                        }
-
-                        if timer.elapsed().as_millis() > 10 {
-                            // We didn't drain the entire queue, so indicate that we
-                            // should run again. We suppress the activation when the
-                            // queue is drained, as in that case librdkafka is
-                            // configured to unpark our thread when a new message
-                            // arrives.
-                            activator.activate();
-                            return SourceStatus::Alive;
-                        }
+            dp_info.refresh_source_information(&mut cp_info, &activator);
+
+            while let Some(message) = dp_info.get_next_message(&cp_info, &activator) {
+                let partition = message.partition;
+                let partition_metrics = &dp_info.partition_metrics.get(&partition).unwrap();
+                let offset = MzOffset::from(message.offset);
+
+                partition_metrics.offset_received.set(offset.offset);
+
+                // Determine the timestamp to which we need to assign this message
+                let ts = find_matching_timestamp(&id, partition, offset, &timestamp_histories);
+                match ts {
+                    None => {
+                        // We have not yet decided on a timestamp for this message,
+                        // we need to buffer the message
+                        dp_info.buffer_message(message);
+                        activator.activate();
+                        return SourceStatus::Alive;
+                    }
+                    Some(ts) => {
+                        // Note: empty and null payload/keys are currently
+                        // treated as the same thing.
+                        let key = message.key.unwrap_or_default();
+                        let out = message.payload.unwrap_or_default();
+                        cp_info
```

Comment:
The unwrap is valid because if we saw an entry with an unknown partition in `downgrade_capability`, we would have set `needs_refresh`, and then the partition metadata would have been inserted in `refresh_source_information` already -- is that right?

Can you add a comment explaining that?

Reply:
Yep.